Source/SGDLib/SimpleDistGradAggregator.h

```diff
@@ -326,36 +305,105 @@ class SimpleDistGradAggregator : public IDistGradAggregator<ElemType>
         if (!m_mpi->IsMainNode())
             m_mpi->Isend(headerCPU, headerCPU->Size(), MPI_CHAR, m_mpi->MainNodeRank(), numGradMatrices, &sendHeaderRequest) || MpiFail("MPI_Isend");
 
-        // Perform async allreduce on the gradient data
+
+        // New aggregation pipeline for non-GDR, perform sync allreduce on the gradient data
+        // For CPU, still use async allreduce
         std::vector<MPI_Request> allReduceRequests;
-        if (!m_nccl.IsSupported())
+        size_t gpuToCpuIndex = 0;
+        size_t cpuToGpuIndex = 0;
+        size_t allReduceIndex = 0;
+        // non-GDR && GPU
+        if ((m_mpi->UseGpuGdr() == 0) && (deviceId != CPUDEVICE))
         {
-            size_t allReduceIndex = 0;
+            Matrix<ElemType>* gpuCopyBuffer = m_aggregationBuffer.get();
+
             ElemType* reductionBuffer;
+            // CurrentGradientIndex will load the index from m_gradientIndexToAggregate
+            size_t currentGradientIndex = -2;  
+            //pipeline loop
             for (size_t i : m_gradientIndexToAggregate)
             {
-                allReduceRequests.push_back(MPI_Request());
-                reductionBuffer = (i == -1)? m_aggregationBuffer->Data() : gradients[i]->Data();
-                if (m_mpi->UseGpuGdr() == 0 && deviceId != CPUDEVICE)
+                if (i != -1)
                 {
-                    m_gpuDataTransferers[allReduceIndex]->WaitForCopyGPUToCPUAsync();
-                    reductionBuffer = m_intermediateCPUBuffers[allReduceIndex].get();
+                    gpuCopyBuffer = gradients[i];
+                }
+                else
+                {
+                    // i == -1, first element is for packed gradients, which should not be with AsyncAggregation
+                    assert(m_useAsyncAggregation == false);
+                }
+                //1.first sync_g_to_c_copy
+                if (gpuToCpuIndex == 0)
```

Comment:
Thank you. I agree that the logic will be more clear. But about "(2) in the loop", I think the order of operations should be "start async copy GPU->CPU for next gradient, wait for last GPU->CPU copy, do aggregation, start async CPU-GPU".  "Wait" shouldn't block the next GPU->CPU copy. What's your opinion?